File: 백준/Gold/2698. 인접한 비트의 개수/인접한 비트의 개수.py
def solve(idx, cnt, bit):
    global dp
    if cnt < 0:
        return 0

    if idx == 1:
        return 1 if cnt == 0 else 0

    if (idx, cnt, bit) in dp:
        return dp[(idx, cnt, bit)]

    ret = solve(idx - 1, cnt, 0)
    if bit == 0:
        ret += solve(idx - 1, cnt, 1)
    else:
        ret += solve(idx - 1, cnt - 1, 1)

    dp[(idx, cnt, bit)] = ret
    return ret


if __name__ == "__main__":
    T = int(input())
    MAX_K = 99
    MAX_N = 100
    dp = dict()
    ans = []
    for _ in range(T):
        N, K = map(int,input().split())


        ans.append(solve(N, K, 0) + solve(N,K,1))

    print('\n'.join(map(str, ans)))

File: 백준/Gold/2698. 인접한 비트의 개수/인접한 비트의 개수.java
import java.util.*;
import java.io.*;

public class Main{
	public static void main(String[] args) throws IOException{
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		T = Integer.parseInt(br.readLine());
		StringBuilder sb = new StringBuilder();
		dp = new int[MAX_N+1][MAX_K+1][2];
		for(int n = 0; n<=MAX_N;n++) {
			for(int k=0; k<=MAX_K;k++) {
				for(int i = 0; i <2 ; i++)
					dp[n][k][i] = -1;
			}
		}
		
		for(int t=0; t<T;t++) {
			st = new StringTokenizer(br.readLine());
			N = Integer.parseInt(st.nextToken());
			K = Integer.parseInt(st.nextToken());
			sb.append(solve(N,K,0) + solve(N,K,1)).append('\n');
		}
		System.out.println(sb);
	}
	
	static int solve(int idx, int cnt ,int bit) {
		if(cnt < 0)
			return 0;
		if(idx == 1) {
			return cnt == 0 ? 1 : 0;
		}
		
		if(dp[idx][cnt][bit] != -1) {
			return dp[idx][cnt][bit];
		}
		
		int ret = solve(idx - 1, cnt, 0);
		if(bit == 0)
			ret += solve(idx-1,cnt,1);
		else
			ret += solve(idx-1,cnt-1,1);
		
		return dp[idx][cnt][bit] = ret;
			
	}
	
	static int MAX_N = 100;
	static int MAX_K = 100;
	static int dp[][][];
	static int T,N,K;
	static StringTokenizer st;
}